### PR TITLE
Remove existing authentication and cart functionality (#2950)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by achave11** [comment]
Unassigning myself and assigning @noah-aviel-dove, as [requested](https://ucsc-gi.slack.com/archives/C705Y6G9Z/p1618263682137400?thread_ts=1618261656.136500&cid=C705Y6G9Z) by 
@hannes-ucsc.

**Review by achave11 — APPROVED** [review]
Great work. Only a weak observation, not sure if the changes would benefit from having multiple commits, like file/path renames and addition/deletions. However, since it is a refactor, it can be justified for the changes to be done in a single commit.

**Comment by amarjandu** [comment]
@achave11 I though about that a little, maybe it looks like having a commit history that looks like:

```
[1/3] Move entire service.app file to the attic
[2/3] Remove Auth/Cart segments out of the lambda/service/app
[3/3] Remove everything not related to the auth/cart stuff from the attic/lamdas/service/app
```

That might be a better way to track all the changes.

@noah-aviel-dove what you think about this?

**Comment by noah-aviel-dove** [comment]
> @noah-aviel-dove what you think about this?

Sounds a bit convoluted. I agree with @achave11 that the mixture of full rename/moves and partial additions/deletions does make the changes a little harder to track, but I think it's even harder to tell what's going on when the same files get touched in multiple commits for what is functionally a single change.

Having a `[1/2]` commit for moving entire files and and `[2/2]` commit for editing the unmoved files, or vice-versa, would be the clearest solution IMO. But a single commit also seems acceptable.

File: attic/lambdas/service/app.py [hunk 1]
```diff
@@ -0,0 +1,704 @@
+@app.authorizer()
```
Comment (noah-aviel-dove) [inline attic/lambdas/service/app.py:1]: I feel this file should preserve the import statements referenced in the moved code, otherwise we're not really preserving it.

Comment (noah-aviel-dove) [reply attic/lambdas/service/app.py:1]: Hard to tell whether this is correct advice or not without going every file in the attic. It just seems to me that there's a huge number of classes referenced in this code here that might be ambiguous to resolve in the future.

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]

File: attic/lambdas/service/app.py (hunk 1 above)
Comment (hannes-ucsc) [reply attic/lambdas/service/app.py:32]: The main reason for the attic is not so much archival but maintaining a list of what used to exist. If you we just deleted files, it would be harder to find the commit that deleted a particular without knowing where it used to exist and what its name was. 

To get to the *exact* original, one can always look at the commit that placed the file into the attic.

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
Adding the import statements to the attic introduced conflicts. In light of @hannes-ucsc's clarifying comments, it seems better to omit the import statements from the attic after all.

**Comment by noah-aviel-dove** [comment]
Additionally, there are several lingering references to the cart functionality throughout the codebase and TF templates:

```
~/PycharmProjects/azul-sc$ grep -i cart $(find src/azul/ -name '*.py')
src/azul/plugins/__init__.py:    cart_item: Mapping[str, Sequence[str]]
src/azul/plugins/metadata/hca/__init__.py:            cart_item={
src/azul/service/lambda_iam_policy.py:                f"arn:aws:dynamodb:{aws.region_name}:{aws.account}:table/{config.dynamo_cart_table_name}",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:dynamodb:{aws.region_name}:{aws.account}:table/{config.dynamo_cart_item_table_name}",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:dynamodb:{aws.region_name}:{aws.account}:table/{config.dynamo_cart_table_name}/index/*",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:dynamodb:{aws.region_name}:{aws.account}:table/{config.dynamo_cart_item_table_name}/index/*",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:states:{aws.region_name}:{aws.account}:stateMachine:{config.cart_item_state_machine_name}",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:states:{aws.region_name}:{aws.account}:stateMachine:{config.cart_export_state_machine_name}"
src/azul/service/lambda_iam_policy.py:                f"arn:aws:states:{aws.region_name}:{aws.account}:execution:{config.cart_item_state_machine_name}:*",
src/azul/service/lambda_iam_policy.py:                f"arn:aws:states:{aws.region_name}:{aws.account}:execution:{config.cart_export_state_machine_name}"
src/azul/service/user_service.py:                                             item={'UserId': user_id, 'DefaultCartId': None})
src/azul/service/user_service.py:    def update(self, user_id: str, default_cart_id):
src/azul/service/user_service.py:        :param default_cart_id: the ID of the default cart (nullable)
src/azul/service/user_service.py:        if default_cart_id is not None:
src/azul/service/user_service.py:            update_conditions['DefaultCartId'] = None
src/azul/service/user_service.py:                                                    update_values={'DefaultCartId': default_cart_id},
src/azul/service/user_service.py:            raise UpdateError(user_id, default_cart_id)
src/azul/service/elasticsearch_service.py:    def transform_cart_item_request(self,
src/azul/service/elasticsearch_service.py:        Create a query using the given filter used for cart item requests
src/azul/service/elasticsearch_service.py:        :param entity_type: type of entities to write to the cart
src/azul/service/elasticsearch_service.py:        source_filter = list(chain(service_config.cart_item['bundles'],
src/azul/service/elasticsearch_service.py:                                   service_config.cart_item[entity_type]))
src/azul/__init__.py:    def dynamo_cart_table_name(self):
src/azul/__init__.py:        return self.qualified_resource_name('carts')
src/azul/__init__.py:    def dynamo_cart_item_table_name(self):
src/azul/__init__.py:        return self.qualified_resource_name('cartitems')
src/azul/__init__.py:    cart_item_write_lambda_basename = 'cartitemwrite'
src/azul/__init__.py:    def cart_item_state_machine_name(self):
src/azul/__init__.py:        return self.qualified_resource_name('cartitems')
src/azul/__init__.py:    def cart_export_max_batch_size(self):
src/azul/__init__.py:        return int(os.environ['AZUL_CART_EXPORT_MAX_BATCH_SIZE'])
src/azul/__init__.py:    def cart_export_min_access_token_ttl(self):
src/azul/__init__.py:        return int(os.environ['AZUL_CART_EXPORT_MIN_ACCESS_TOKEN_TTL'])
src/azul/__init__.py:    def cart_export_state_machine_name(self):
src/azul/__init__.py:        return self.qualified_resource_name('cartexport')
src/azul/__init__.py:    cart_export_dss_push_lambda_basename = 'cartexportpush'
```

It's not clear to me whether these fall under the scope of this ticket, since the title only mentioned the "cart service". But the (undocumented) `UserService` class and its associated unit tests only seem to be used by the attic-ed `CartItemManager`, so perhaps they ought to be attic-ed as well. Same for `ElasticSearchService.transform_cart_item_request`. I recommend discussing this with @hannes-ucsc at parking lot, unless he weighs in before then.

**Comment by hannes-ucsc** [comment]
The ticket speaks of "cart functionality", so yes, that stuff should be moved to attic as well. Consider collecting everything in a sub-directory of the attic e.g., `attic/cart`.

**Comment by amarjandu** [comment]
Confirmed IT passes on personal deployment.

**Comment by noah-aviel-dove** [comment]
Diff looks good. I'm curious as to why the upgrading instruction are necessary.

**Review by noah-aviel-dove — APPROVED** [review]

**Comment by amarjandu** [comment]
@noah-aviel-dove I removed the @app.authorizer from the service.app, this messes with the API Gateway configuration, Terraform is not able to handle the teardown by itself, it gets stuck into a cyclical dependency. Nuking the API Gateway instance for the service lambda ensures that on the next deploy the Auth infra is able to be torn down.

I had the `Error: Cycle `saved in the terminal scrollback of my buffer so I could show you a screenshot , but I used `command+k` to clear my scrollback buffer earlier today >.> ( I usually use `clear`) https://superuser.com/questions/819593/what-is-the-difference-between-ctrll-and-cmdk-for-clearing-the-terminal-screen

Anyway, yeah that's why there are upgrading instructions here.

File: UPGRADING.rst [hunk 2]
```diff
@@ -11,6 +11,23 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#2950 Move auth and cart service to attic
+=========================================
+
+1. Before upgrading to this commit, run ::
```
Comment (hannes-ucsc) [inline UPGRADING.rst:17]: Why is this necessary?

File: attic/security/retorts.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
+from base64 import (
```
Comment (hannes-ucsc) [inline attic/security/retorts.py:1]: This file should not be in the attic since it is still present in `src`. Just deleting the unused artifacts from the copy in `src` should suffice.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: UPGRADING.rst (hunk 2 above)
Comment (amarjandu) [reply UPGRADING.rst:17]: Terraform cant figure out how to properly teardown the infra due to a `Cycle Error`:

Screen Shot 2021-04-15 at 11 17 14 AM
https://github.com/DataBiosphere/azul/pull/2969#issuecomment-819896848

Comment (hannes-ucsc) [reply UPGRADING.rst:17]: OK, I've tried my mightiest yesterday to avoid this workaround, to no avail.

Comment (hannes-ucsc) [reply UPGRADING.rst:17]: https://github.com/DataBiosphere/azul/issues/2992

**Review by hannes-ucsc — APPROVED** [review]

**Comment by amarjandu** [comment]
Force pushed branch after approval  due to force push to `develop` branch.

**Comment by achave11** [comment]
Unit test not passing in `sandbox`

**Comment by amarjandu** [comment]
unable to replicate travis failure locally.
ran `git prsq` then pushed re-based changes to this PR branch.
tests now show passing.
